File: data_structures/stack/C/stack.c
#include <stdio.h>
#include <stdlib.h>

#define max 10
int top = -1;
int stack[max];

void push();
void pop();
void isEmpty();
void isFull();
void peek();
void count();
void change();
void display();

void push() // Insert element at top
{
    if (top == max - 1)
    {
        printf("Stack overflow\n");
    }
    else
    {
        int val;
        printf("Enter value to be inserted: ");
        scanf("%d", &val);
        top += 1;
        stack[top] = val;
    }
}

void pop() // Remove element from top
{
    if (top == -1)
    {
        printf("Stack underflow\n");
    }
    else
    {
        printf("Popped element: %d\n", stack[top]);
        top -= 1;
    }
}

void isEmpty() // Check if stack is empty
{
    if (top == -1)
    {
        printf("Stack is empty\n");
    }
    else
    {
        printf("Stack is not empty\n");
    }
}

void isFull() // Check if stack is full
{
    if (top == max - 1)
    {
        printf("Stack is full\n");
    }
    else
    {
        printf("Stack is not full\n");
    }
}

void peek() // View top element
{
    if (top == -1)
    {
        printf("Stack is empty\n");
    }
    else
    {
        printf("Top element: %d\n", stack[top]);
    }
}

void count() // Count number of elements in stack
{
    printf("Number of elements in stack: %d\n", top + 1);
}

void change() // Change element at specific position
{
    int pos, newVal;
    printf("Enter position to change: ");
    scanf("%d", &pos);

    if (pos >= 0 && pos <= top)
    {
        printf("Enter new value: ");
        scanf("%d", &newVal);
        stack[pos] = newVal;
        printf("Value changed at position %d\n", pos);
    }
    else
    {
        printf("Invalid position\n");
    }
}

void display() // Display all stack elements
{
    if (top == -1)
    {
        printf("Stack is empty\n");
    }
    else
    {
        printf("Stack elements (top to bottom): ");
        for (int i = top; i >= 0; i--)
        {
            printf("%d ", stack[i]);
        }
        printf("\n");
    }
}

int main() // Main function with menu
{
    int ch;
    do
    {
        printf("\n1) Push\n2) Pop\n3) Peek\n4) Count\n5) Change\n6) Display\n7) Is full\n8) Is empty\n9) Exit\nEnter choice: ");
        scanf("%d", &ch);

        switch (ch)
        {
        case 1:
            push();
            break;

        case 2:
            pop();
            break;

        case 3:
            peek();
            break;

        case 4:
            count();
            break;

        case 5:
            change();
            break;

        case 6:
            display();
            break;

        case 7:
            isFull();
            break;

        case 8:
            isEmpty();
            break;
        case 9:
            exit(0);
            break;

        default:
            printf("Invalid choice\n");
            break;
        }
    } while (ch != 9);

    return 0;
}

File: data_structures/Stack/java/stack.java
import java.util.Scanner;

/**
 * Stack implementation in Java using array.
 *
 * Operations supported:
 * 1. Push
 * 2. Pop
 * 3. Peek
 * 4. Display
 *
 * Time Complexity:
 * Push, Pop, Peek: O(1)
 * Display: O(n)
 *
 * Space Complexity: O(n)
 */
public class stack {
    private int maxSize;
    private int[] stackArray;
    private int top;

    // Constructor to initialize stack
    public stack(int size) {
        maxSize = size;
        stackArray = new int[maxSize];
        top = -1;
    }

    // Push element onto stack
    public void push(int value) {
        if (top >= maxSize - 1) {
            System.out.println("Stack Overflow!");
        } else {
            stackArray[++top] = value;
            System.out.println(value + " pushed onto stack.");
        }
    }

    // Pop element from stack
    public void pop() {
        if (top < 0) {
            System.out.println("Stack Underflow!");
        } else {
            int value = stackArray[top--];
            System.out.println(value + " popped from stack.");
        }
    }

    // Peek top element
    public void peek() {
        if (top < 0) {
            System.out.println("Stack is empty!");
        } else {
            System.out.println("Top element is " + stackArray[top]);
        }
    }

    // Display stack elements
    public void display() {
        if (top < 0) {
            System.out.println("Stack is empty!");
        } else {
            System.out.println("Stack elements:");
            for (int i = top; i >= 0; i--) {
                System.out.println(stackArray[i]);
            }
        }
    }

    // Main method for dynamic input
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter stack size: ");
        int size = sc.nextInt();
        stack stack = new stack(size);

        while (true) {
            System.out.println("\nChoose operation: 1-Push 2-Pop 3-Peek 4-Display 5-Exit");
            int choice = sc.nextInt();
            switch (choice) {
                case 1:
                    System.out.print("Enter value to push: ");
                    int value = sc.nextInt();
                    stack.push(value);
                    break;
                case 2:
                    stack.pop();
                    break;
                case 3:
                    stack.peek();
                    break;
                case 4:
                    stack.display();
                    break;
                case 5:
                    System.out.println("Exiting...");
                    sc.close();
                    System.exit(0);
                default:
                    System.out.println("Invalid choice!");
            }
        }
    }
}
